plugins/inputs/device/device.go

```diff
@@ -0,0 +1,167 @@
+package device
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type Device struct {
+	Type    string   `toml:"type"`
+	Devices []string `toml:"devices"`
+
+	devicetypes map[string]DeviceDescriptor
+}
+
+type DeviceDescriptor struct {
+	Fields []DeviceField
+}
+
+type DeviceField struct {
+	Path   string
+	Name   string
+	Type   DeviceFieldType
+	Factor float64
+}
+
+type DeviceFieldType int
+
+const (
+	DeviceFieldType_Integer DeviceFieldType = iota + 1
+	DeviceFieldType_Float
+	DeviceFieldType_String
+	DeviceFieldType_Bool
+)
+
+const sampleConfig = `
+  # device type, only specified once
+  type = "bme280"
+  # type = "ina219"
+
+  # list of device paths
+  devices = ["/sys/bus/i2c/devices/1-0076/iio:device0"]
+`
+
+// SampleConfig returns the default configuration of the Input
+func (d *Device) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Device) Description() string {
+	return "Collect metrics from devices not supported by lm_sensors"
+}
+
+func (d *Device) Gather(acc telegraf.Accumulator) error {
+	devicetype, ok := d.devicetypes[d.Type]
+
+	if !ok {
+		return errors.New(fmt.Sprintf("unknown device type %s", d.Type))
+	}
+
+	now := time.Now()
+
+	for _, devpath := range d.Devices {
+		tags := map[string]string{
+			"device": devpath,
+			"type":   d.Type,
+		}
+
+		fields := make(map[string]interface{})
+
+		for _, field := range devicetype.Fields {
+			fileContents, err := ioutil.ReadFile(path.Join(devpath, field.Path))
+
+			if err != nil {
+				return err
+			}
+
+			vStr := string(bytes.TrimSpace(bytes.Trim(fileContents, "\x00")))
+
+			var value interface{}
+
+			switch field.Type {
+			case DeviceFieldType_Integer:
+				value, err = strconv.Atoi(vStr)
+			case DeviceFieldType_Float:
+				value, err = strconv.ParseFloat(vStr, 64)
+			case DeviceFieldType_String:
+				value = vStr
+			case DeviceFieldType_Bool:
+				value, err = strconv.ParseBool(vStr)
+			}
+
+			if err != nil {
+				return err
+			}
+
+			if field.Type == DeviceFieldType_Float {
+				value = value.(float64) * field.Factor
+			}
+
+			fields[field.Name] = value
+		}
+
+		acc.AddGauge("device", fields, tags, now)
+	}
+	return nil
+}
+
+func init() {
+	types := map[string]DeviceDescriptor{
```

Comment:
I don't think you should insert this default devices as not every people who use this has ina219 and bme280.

CMIIW.

Reply:
I plan to add either more of these predefined devices (these two are the only ones I have on hand to test) or make the device descriptor completely customizable via telegraf.conf. This list is not at all exhaustive.
I'm not sure which option would fit better to the overall project, maybe it would also be better to move the type definitions to a new file to not clutter the actual code.

Reply:
I personally prefer to have a customizable device descriptors (just provide a commented one for example in the sample config) rather than a list of predefined ones. 

As for the unit test, I think you should make mock device and look at other input plugins as for how it should be tested.

Reply:
Your're right that's probably the better solution.

How would I make a mock device? Just provide a directory with some regular text files like in [the file plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/file/dev/testfiles)? Quick question: the docker-compose.yml in that folder has nothing to do with the go tests, are both required?

Reply:
If that's how your plugin works, then you can do that.

The docker-compose.yml from input file plugin is clearly just to test locally judging by the directory's name `dev`. I don't know why he put it there though.